File: android/src/main/kotlin/com/pravera/flutter_activity_recognition/StreamHandlerImpl.kt
package com.pravera.flutter_activity_recognition

import android.content.Context
import com.pravera.flutter_activity_recognition.errors.ErrorCodes
import com.pravera.flutter_activity_recognition.service.ActivityDataCallback
import com.pravera.flutter_activity_recognition.service.ActivityRecognitionManager
import com.pravera.flutter_activity_recognition.utils.ErrorHandleUtils
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

/** StreamHandlerImpl */
class StreamHandlerImpl(private val context: Context) : EventChannel.StreamHandler {
	private lateinit var eventChannel: EventChannel
	private lateinit var activityRecognitionManager: ActivityRecognitionManager

	private var binding: ActivityPluginBinding? = null

	fun init(messenger: BinaryMessenger) {
		activityRecognitionManager = ActivityRecognitionManager()
		eventChannel = EventChannel(messenger, "flutter_activity_recognition/updates")
		eventChannel.setStreamHandler(this)
	}

	fun dispose() {
		if (::eventChannel.isInitialized) {
			eventChannel.setStreamHandler(null)
		}
	}

	fun onAttachedToActivity(binding: ActivityPluginBinding) {
		this.binding = binding
	}

	fun onDetachedFromActivity() {
		binding = null
	}

	override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
		activityRecognitionManager.startService(
			context = context,
			callback = object : ActivityDataCallback {
				override fun onUpdate(activityJson: String) {
					events.success(activityJson)
				}

				override fun onError(errorCode: ErrorCodes) {
					ErrorHandleUtils.handleMethodCallError(events, errorCode)
				}
			}
		)
	}

	override fun onCancel(arguments: Any?) {
		activityRecognitionManager.stopService(context)
	}
}

File: ios/Classes/StreamHandlerImpl.swift
//
//  StreamHandlerImpl.swift
//  flutter_activity_recognition
//
//  Created by WOO JIN HWANG on 2021/04/23.
//

import Foundation

class StreamHandlerImpl : NSObject, FlutterStreamHandler {
  private let eventChannel: FlutterEventChannel
  private let activityRecognitionManager: ActivityRecognitionManager
  
  init(messenger: FlutterBinaryMessenger) {
    self.activityRecognitionManager = ActivityRecognitionManager()
    self.eventChannel = FlutterEventChannel(name: "flutter_activity_recognition/updates", binaryMessenger: messenger)
    super.init()
    self.eventChannel.setStreamHandler(self)
  }
  
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    let handler = ActivityDataHandlerImpl(events)
    activityRecognitionManager.startService(handler: handler)
    return nil
  }
  
  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    activityRecognitionManager.stopService()
    return nil
  }
}
